feat(react): add useErrorBoundaryFallbackProps (#216)
suspensive/react: packages/react/src/ErrorBoundary.en.mdx

![Example banner](/gif/errorboundary-example.gif)

:::tip

## Use `<ErrorBoundary/>` fallback props without prop drilling

> ### useErrorBoundaryFallbackProps

Reviewer: I think we need to seperate component's doc and hook's doc.

I think you intended combine this hook with component document, because it is related to this component.
But I think it is more convenient to seperated component document  with hook document

reference is below
<img width="401" alt="image" src="https://github.com/suspensive/react/assets/57122180/843de792-cece-4f1e-b1f2-83ba393960e4">

Chakra's intend is like that,
> If you need to use hook, check this [link](https://chakra-ui.com/docs/components/checkbox/usage#hooks)

<img width="1143" alt="image" src="https://github.com/suspensive/react/assets/57122180/b990e870-a0ff-435a-b6df-0031ce38efb6">

But our document is mixed.

Author: I made https://github.com/suspensive/react/issues/220 to resolve it. Let's talk about it after this Pull Request

In fallback of `<ErrorBoundary/>`, by calling useErrorBoundaryFallbackProps().reset/error, we can use error of `<ErrorBoundary/>` and reset it without using prop directly

```tsx
import { ErrorBoundary, useErrorBoundaryFallbackProps } from '@suspensive/react'

const Nested = () => {
  const props = useErrorBoundaryFallbackProps()

  return (
    <>

>       <button onClick={props.reset} />

Reviewer: ```suggestion
      <button onClick={props.reset}>click to reset error!</button>
```

how about like this?

Author: I updated it in [1e748f8](https://github.com/suspensive/react/pull/216/commits/1e748f837c19d8e3cd4214e363fece8e62ddfc6e)

Almost cases we use text as `Try again` to reset ErrorBoundary

> ```tsx
> import { ErrorBoundary, useErrorBoundaryFallbackProps } from '@suspensive/react'
> 
> const Nested = () => {
>   const props = useErrorBoundaryFallbackProps()
> 
>   return (
>     <>
>       <button onClick={props.reset}>Try again</button>
>       {props.error.message}
>     </>
>   )
> }

Reviewer 2: ```suggestion
```tsx
import { ErrorBoundary, useErrorBoundaryFallbackProps } from '@suspensive/react'

const Nested = () => {
  const { reset, error } = useErrorBoundaryFallbackProps()

  return (
    <>
      <button onClick={reset}>Try again</button>
      {error.message}
    </>
  )
}
```

> const ErrorBoundaryFallback = () => <Nested />
> const Example = () => (
>   <ErrorBoundary fallback={ErrorBoundaryFallback}>

Reviewer: ```suggestion
const Example = () => (
  <ErrorBoundary fallback={Nested}>
```

What about this?
I think `ErrorBoundaryFallback` component is not necessary
or, Nested component rename to ErrorBoundaryFallback

Reviewer: Oh you intend to props drilling case... I miss understand sorry

>   <ErrorBoundary fallback={ErrorBoundaryFallback}>

Reviewer: How about add onReset prop ??

User who show this example or document first time, just draged whole example and paste their code. and curious "why my query is not reseted and don't query again?"

Author: > and curious "why my query is not reseted and don't query again?"

I guess `query` you mentioned is for @tanstack/react-query. but because @suspensive/react is not only for @tanstack/react-query, I think we should not to describe this example with @tanstack/react-query

Reviewer: I agree, thanks!

![Example banner](/gif/errorboundary-example.gif)

:::tip

## Define component as `<ErrorBoundary/>`'s fallback

### ErrorBoundaryFallbackProps

If you want to deliver a declared component as `<ErrorBoundary/>`'s fallback, you can use the `ErrorBoundaryFallbackProps` type to declare the component easily.

```tsx

> import type { ErrorBoundaryFallbackProps } from '@suspensive/react'
> 
> const ErrorBoundaryFallback = ({ reset, error }: ErrorBoundaryFallbackProps) => (
>   <>
>     <button onClick={reset}>reset</button>
>     {error.message}
>   </>
> )
> 
> const Example = () => (
>   <ErrorBoundary fallback={ErrorBoundaryFallback}>
>     <ErrorAfter4s />
>   </ErrorBoundary>
> )

Reviewer 3: If it's for the purpose of providing typing

Wouldn't it be better for the developer experience to define not only the type of props but also the return?

```suggestion
import type { ErrorBoundaryFallbackType } from '@suspensive/react'

const ErrorBoundaryFallback: ErrorBoundaryFallbackType = ({ reset, error }) => (
  <>  // The type of return value is also certain
    <button onClick={reset}>reset</button>
    {error.message}
  </>
)

const Example = () => (
  <ErrorBoundary fallback={ErrorBoundaryFallback}> 
    <ErrorAfter4s />
  </ErrorBoundary>
)
```

I inspried by nextjs GetServerSideProps

Author: > Wouldn't it be better for the developer experience to define not only the type of props but also the return?

I think that's a good idea too. However, even if we do not provide the return type, TypeScript will create a type error if it does not match the fallback type of ErrorBoundary.

I think we need to consider whether providing a return type is right choice. Because the return type may be different from the actual thing returned ([this video](https://www.youtube.com/watch?v=I6V2FkW1ozQ&ab_channel=Theo-t3%E2%80%A4gg) explain it well). If it is as essential as ErrorBoundaryFallbackProps, it would be a good idea to add it as a minor version feature.

Let's consider it after merging this
